expression/extra/result/catch.py

```diff
@@ -0,0 +1,51 @@
+from functools import wraps
+from typing import Any, Callable, Optional, Type, TypeVar, Union, cast, overload
+
+from expression.core import Error, Ok, Result
+
+TSource = TypeVar("TSource")
+TError = TypeVar("TError", bound=Exception)
+TError_ = TypeVar("TError_", bound=Exception)
+
+
+@overload
+def catch(
+    exception: Type[TError_],
+) -> Callable[
+    [Callable[..., Union[TSource, Result[TSource, TError]]]], Callable[..., Result[TSource, Union[TError, TError_]]]
+]:
+    ...
+
+
+@overload
+def catch(f: Callable[..., TSource], *, exception: Type[TError]) -> Callable[..., Result[TSource, TError]]:
+    ...
+
+
+def catch(
+    f: Optional[Callable[..., TSource]] = None, *, exception: Type[TError]
+) -> Callable[[Callable[..., TSource]], Union[Callable[..., Result[TSource, TError]], Result[TSource, TError]]]:
+    def decorator(fn: Callable[..., TSource]) -> Callable[..., Result[TSource, TError]]:
+        @wraps(fn)
+        def wrapper(*args: Any, **kwargs: Any) -> Result[TSource, TError]:
+            try:
+                out = fn(*args, **kwargs)
+            except Exception as exn:
+                if isinstance(exn, exception):
+                    return Error(cast("TError", exn))
+                raise
+            else:
+                if isinstance(out, Result):
+                    return cast(Result[TSource, TError], out)
+
+                return Ok(out)
+
+        return wrapper
+
+    if f is not None:
+        return decorator(f)
```

Comment:
This will fail since it is missing the required positional argument `exception`. Letting the first argument of your decorator be a function/method only makes sense if you want to be able to use the decorator without arguments e.g.
```
@catch
def foo():
   ...
```

But it doesn't really make sense to support this here, since we always want to specify an exception to catch. My suggestion would be to implement the decorator like this instead:
 
 ```
def _validate(input: Tuple[Any, ...]) -> None:
    if len(input) == 0:
        raise ValueError("At least one exception must be passed to catch")
    if not all(isinstance(el, type) and Exception in el.__bases__ for el in input):
        raise TypeError("Only exception types can be passed as arguments to catch")

def catch(*exception: Type[TError]) -> Callable[[Callable[..., TSource]], Callable[..., Result[TSource, TError]]]:
    _validate(exception)
    def decorator(fn: Callable[..., TSource]) -> Callable[..., Result[TSource, TError]]:
        @wraps(fn)
        def wrapper(*args: Any, **kwargs: Any) -> Result[TSource, TError]:
            try:
                out = fn(*args, **kwargs)
            except exception as exn:
                return Error(cast("TError", exn))
            else:
                if isinstance(out, Result):
                    return cast(Result[TSource, TError], out)

                return Ok(out)

        return wrapper

    return decorator
 ```
No overloading and disallows using `@catch` without specifying exceptions. It also has the added benefit of letting the user specify an arbitrary number of exception type arguments without chaining decorators.

Usage example:
```

```

Reply:
Could add some runtime validation on the `exception` parameter to give better error messages if the user attempts to use the decorator without arguments.

Reply:
Could also support arbitrary exception arguments by packing them (i.e. `def catch(*exception: Type[TError])` to support catching multiple exceptions without chaining decorators.